File: Java/QuickSort.java
package Java;

public class QuickSort extends Sort{
	
	public static void main(String[] args){
		int[] data = buildDataSet(5);
		
		timeSort(new QuickSort(), data);
	}
	
	public QuickSort(){ super("Quick"); }
		
	private int medianOfThree(int[] data, int low, int high){
        int middle = ( low + high ) / 2;
        
        if( data[middle] < data[low] )
            swap(data, low, middle);
        if( data[high] <data[low] )
            swap(data, low, high);
        if( data[high] < data[middle] )
            swap(data, middle, high);

        swap(data, middle, high-1);  

        return high-1;
	}
	
	private int partition(int[] data, int low, int high){
		int pivot = medianOfThree(data, low, high);
		
		//move elements larger than the pivot
		//to the right of the pivot
		while (low < pivot)
		{
			if ( data[low] > data[pivot] )
			{
				int temp = data[low];
				data[low] = data[pivot - 1];
				data[pivot - 1] = data[pivot];
				data[pivot] = temp;
				pivot--;
			}
			else
				low++;
		}
		return pivot;
	}
	
	private void quickSort(int[] data, int low, int high){
		if ( low < high)
		{
			int index = partition(data, low, high);
			quickSort(data, low, index - 1);
			quickSort(data, index + 1, high);
		}
	}
	
	@Override
	public void sort(int[] data){
		quickSort(data, 0, data.length - 1);
	}
	
	private void swap(int[] data, int index1, int index2){
		int temp = data[index1];
		data[index1] = data[index2];
		data[index2] = temp;
	}
}

File: javascript/quick-sort.js
function swap(array, index1, index2){
    var temp = array[index1]
    array[index1] = array[index2]
    array[index2] = temp
}

function medianOfThree(array, low, high){
    var middle = (high - low) / 2 + low

    if ( array[low] > array[middle] )
        swap(array, low, middle)
    if ( array[low] > array[high] )
        swap(array, low, high)
    if ( array[middle] > array[high] )
        swap(array, middle, high)
    
    swap(array, middle, high-1)

    return high-1
}

function partition(array, low, high){
    var pivot = medianOfThree(array, low, high)

    while ( low < pivot )
    {
        if ( array[low] > array[pivot] )
        {
            let temp = array[low]
            array[low] = array[pivot-1]
            array[pivot-1] = array[pivot]
            array[pivot--] = temp
        }
        else
            low++
    }
    return pivot
}

function quickSort(array, low, high){
    if ( low < high )
    {
        let index = partition(array, low, high)
        quickSort(array, low, index - 1)
        quickSort(array, index + 1, high)
    }
}

function sort(array){
    quickSort(array, 0, array.length - 1)
}

